<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/Images/SN_logo.png" width="300" alt="cognitiveclass.ai logo">
</center>


# Extracting and Visualizing Stock Data
## Description

Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working in IBM Cloud Watson Studio, please replace the command for installing nbformat from `!pip install nbformat==4.2.0` to simply `!pip install nbformat`


In [2]:

!python -m pip install yfinance
!python -m pip install pandas
!python -m pip install requests
!python -m pip install bs4
!python -m pip install plotly

  Using cached yfinance-0.2.31-py2.py3-none-any.whl (65 kB)
                                              0.0/10.6 MB ? eta -:--:--
                                              0.0/10.6 MB ? eta -:--:--
                                              0.0/10.6 MB ? eta -:--:--
                                              0.0/10.6 MB ? eta -:--:--
                                             0.0/10.6 MB 145.2 kB/s eta 0:01:14
                                             0.0/10.6 MB 145.2 kB/s eta 0:01:14
                                             0.0/10.6 MB 150.6 kB/s eta 0:01:11
                                             0.1/10.6 MB 217.9 kB/s eta 0:00:49
                                             0.1/10.6 MB 290.5 kB/s eta 0:00:37
                                             0.2/10.6 MB 535.8 kB/s eta 0:00:20
     -                                       0.3/10.6 MB 680.9 kB/s eta 0:00:16
     -                                        0.5/10.6 MB 1.1 MB/s eta 0:00:10
     --      

  Using cached bs4-0.0.1.tar.gz (1.1 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Created wheel for bs4: filename=bs4-0.0.1-py3-none-any.whl size=1264 sha256=96a834fed19331df1234f956ff01e6c74301edc9c08aa422035d2a2e7b24f91c
  Stored in directory: c:\users\casper\appdata\local\pip\cache\wheels\d4\c8\5b\b5be9c20e5e4503d04a6eac8a3cd5c2393505c29f02bea0960
Successfully built bs4
  Using cached plotly-5.17.0-py2.py3-none-any.whl (15.6 MB)
  Using cached tenacity-8.2.3-py3-none-any.whl (24 kB)


In [3]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [4]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [5]:
tesla = yf.Ticker("TSLA")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [6]:
tesla_data = tesla.history("max")

**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [7]:
tesla_data.reset_index(inplace=True)

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [32]:
url = " https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"
html_data = requests.get(url).text

Parse the html data using `beautiful_soup`.


In [33]:
soup = BeautifulSoup(html_data, "html5lib")
soup

<!DOCTYPE html>
<!--[if lt IE 7]>      <html class="no-js lt-ie9 lt-ie8 lt-ie7"> <![endif]--><!--[if IE 7]>         <html class="no-js lt-ie9 lt-ie8"> <![endif]--><!--[if IE 8]>         <html class="no-js lt-ie9"> <![endif]--><!--[if gt IE 8]><!--><html class="no-js"><!--<![endif]--><head>
        <meta charset="utf-8"/>
        <meta content="IE=edge,chrome=1" http-equiv="X-UA-Compatible"/>
		<link href="https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue" rel="canonical"/>
		<title>Tesla Revenue 2010-2022 | TSLA | MacroTrends</title>
		<meta content="Tesla annual/quarterly revenue history and growth rate from 2010 to 2022. Revenue can be defined as the amount of money a company receives from its customers in exchange for the sales of goods or services.  Revenue is the top line item on an income statement from which all costs and expenses are subtracted to arrive at net income.    
				
				&lt;ul style='margin-top:10px;'&gt;
				&lt;li&gt;Tesla revenue for the quarter ending

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [68]:
tesla_revenue = {"Date" : None}
            
table_rows = soup.find_all("tbody")

for i, row in enumerate(table_rows):
    
    if i == 1:
        print("row", i)
        cells = row.find_all("td")

        for j, cell in enumerate(cells):
            print("column", j, "cell", cell.string)
            
            if j % 2 == 0:
                
                tesla_revenue[cell.string] = ''
                

row 1
column 0 cell 2022-09-30
column 1 cell $21,454
column 2 cell 2022-06-30
column 3 cell $16,934
column 4 cell 2022-03-31
column 5 cell $18,756
column 6 cell 2021-12-31
column 7 cell $17,719
column 8 cell 2021-09-30
column 9 cell $13,757
column 10 cell 2021-06-30
column 11 cell $11,958
column 12 cell 2021-03-31
column 13 cell $10,389
column 14 cell 2020-12-31
column 15 cell $10,744
column 16 cell 2020-09-30
column 17 cell $8,771
column 18 cell 2020-06-30
column 19 cell $6,036
column 20 cell 2020-03-31
column 21 cell $5,985
column 22 cell 2019-12-31
column 23 cell $7,384
column 24 cell 2019-09-30
column 25 cell $6,303
column 26 cell 2019-06-30
column 27 cell $6,350
column 28 cell 2019-03-31
column 29 cell $4,541
column 30 cell 2018-12-31
column 31 cell $7,226
column 32 cell 2018-09-30
column 33 cell $6,824
column 34 cell 2018-06-30
column 35 cell $4,002
column 36 cell 2018-03-31
column 37 cell $3,409
column 38 cell 2017-12-31
column 39 cell $3,288
column 40 cell 2017-09-30
column 41 

In [69]:
dollar_list = ["Revenue"]



for i, row in enumerate(table_rows):
    
    if i == 1:
        cells = row.find_all("td")

        for j, cell in enumerate(cells):
            if j % 2 != 0:
                
                dollar_list.append(cell.string)

In [70]:
dollar_list

['Revenue',
 '$21,454',
 '$16,934',
 '$18,756',
 '$17,719',
 '$13,757',
 '$11,958',
 '$10,389',
 '$10,744',
 '$8,771',
 '$6,036',
 '$5,985',
 '$7,384',
 '$6,303',
 '$6,350',
 '$4,541',
 '$7,226',
 '$6,824',
 '$4,002',
 '$3,409',
 '$3,288',
 '$2,985',
 '$2,790',
 '$2,696',
 '$2,285',
 '$2,298',
 '$1,270',
 '$1,147',
 '$1,214',
 '$937',
 '$955',
 '$940',
 '$957',
 '$852',
 '$769',
 '$621',
 '$615',
 '$431',
 '$405',
 '$562',
 '$306',
 '$50',
 '$27',
 '$30',
 '$39',
 '$58',
 '$58',
 '$49',
 '$36',
 '$31',
 '$28',
 '$21',
 None,
 '$46',
 '$27']

In [71]:
for key in tesla_revenue:
    
    tesla_revenue[key] = dollar_list.pop(0)

In [72]:
tesla_revenue

{'Date': 'Revenue',
 '2022-09-30': '$21,454',
 '2022-06-30': '$16,934',
 '2022-03-31': '$18,756',
 '2021-12-31': '$17,719',
 '2021-09-30': '$13,757',
 '2021-06-30': '$11,958',
 '2021-03-31': '$10,389',
 '2020-12-31': '$10,744',
 '2020-09-30': '$8,771',
 '2020-06-30': '$6,036',
 '2020-03-31': '$5,985',
 '2019-12-31': '$7,384',
 '2019-09-30': '$6,303',
 '2019-06-30': '$6,350',
 '2019-03-31': '$4,541',
 '2018-12-31': '$7,226',
 '2018-09-30': '$6,824',
 '2018-06-30': '$4,002',
 '2018-03-31': '$3,409',
 '2017-12-31': '$3,288',
 '2017-09-30': '$2,985',
 '2017-06-30': '$2,790',
 '2017-03-31': '$2,696',
 '2016-12-31': '$2,285',
 '2016-09-30': '$2,298',
 '2016-06-30': '$1,270',
 '2016-03-31': '$1,147',
 '2015-12-31': '$1,214',
 '2015-09-30': '$937',
 '2015-06-30': '$955',
 '2015-03-31': '$940',
 '2014-12-31': '$957',
 '2014-09-30': '$852',
 '2014-06-30': '$769',
 '2014-03-31': '$621',
 '2013-12-31': '$615',
 '2013-09-30': '$431',
 '2013-06-30': '$405',
 '2013-03-31': '$562',
 '2012-12-31': '$30

Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [74]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")

KeyError: 'Revenue'

Execute the following lines to remove an null or empty strings in the Revenue column.


In [29]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [30]:
tesla_revenue.tail(5)

Empty DataFrame
Columns: [Date, Revenue]
Index: []

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [30]:
gme = yf.Ticker('GME')

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [31]:
gme_data = gme.history(period = "max")

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [32]:
gme_data.reset_index(inplace=True)
gme_data.head(5)

Date      Open      High       Low     Close    Volume  Dividends  \
0 2002-02-13  1.620129  1.693350  1.603296  1.691667  76216000        0.0   
1 2002-02-14  1.712707  1.716074  1.670626  1.683250  11021600        0.0   
2 2002-02-15  1.683251  1.687459  1.658002  1.674834   8389600        0.0   
3 2002-02-19  1.666418  1.666418  1.578047  1.607504   7410400        0.0   
4 2002-02-20  1.615920  1.662210  1.603296  1.662210   6892800        0.0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [36]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
html_data = requests.get(url).text

Parse the html data using `beautiful_soup`.


In [39]:
soup = BeautifulSoup(html_data, "lxml")

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [40]:
gme_revenue = pd.DataFrame(columns = ["Date","Revenue"])

for table in soup.find_all('table'):
    if table.find('th').getText().startswith("GameStop Quarterly Revenue"):
        for row in table.find("tbody").find_all("tr"):
            col = row.find_all("td")
            if len(col) != 2: continue
            Date = col[0].text
            Revenue = col[1].text.replace("$","").replace(",","")
               
            gme_revenue = gme_revenue.append({"Date":Date, "Revenue":Revenue}, ignore_index=True)

AttributeError: 'DataFrame' object has no attribute 'append'

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [41]:
gme_revenue.tail(5)

Empty DataFrame
Columns: [Date, Revenue]
Index: []

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
